### ISSUE TITLE
Spring Boot Hazelcast AutoConfiguration conflicts with Jet

### ISSUE THREAD

**Issue by jerrinot** [hazelcast/hazelcast-jet#1835]
Spring Boot automatically creates a Hazelcast instance when either `Config` bean exist or when it finds a [configuration file](https://github.com/spring-projects/spring-boot/blob/a66c4d3096877b33119366ea9ea165c892b462c8/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastServerConfiguration.java#L78-L79). 

This makes Hazelcast Jet hard to use in the Spring Boot container: Spring will try to create a new IMDG instance when it detects hazelcast.xml (or any other config file), but you probably want to start Jet instance. 

It's actually even worst when you explicitly create a new Jet instance then it will try to create join with the IMDG instance created by Spring and this will fail with:
```
com.hazelcast.core.HazelcastException: Service with name 'hz:impl:jetService' not found!
	at com.hazelcast.spi.impl.NodeEngineImpl.getService(NodeEngineImpl.java:367) ~[hazelcast-jet-3.2.1.jar:3.2.1]
	at com.hazelcast.spi.Operation.getService(Operation.java:395) ~[hazelcast-jet-3.2.1.jar:3.2.1]
	at com.hazelcast.jet.impl.operation.GetClusterMetadataOperation.run(GetClusterMetadataOperation.java:41) ~[hazelcast-jet-3.2.1.jar:3.2.1]
	at com.hazelcast.spi.Operation.call(Operation.java:170) ~[hazelcast-jet-3.2.1.jar:3.2.1]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:210) [hazelcast-jet-3.2.1.jar:3.2.1]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:199) [hazelcast-jet-3.2.1.jar:3.2.1]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:416) [hazelcast-jet-3.2.1.jar:3.2.1]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:153) [hazelcast-jet-3.2.1.jar:3.2.1]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:123) [hazelcast-jet-3.2.1.jar:3.2.1]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:110) [hazelcast-jet-3.2.1.jar:3.2.1]
Caused by: com.hazelcast.spi.exception.ServiceNotFoundException: Service with name 'hz:impl:jetService' not found!
	... 10 common frames omitted
```

The right solution is probably changing Spring Boot autoconfig to not create IMDG instances when Jet classes are detected. This is a good [entry point](https://github.com/spring-projects/spring-boot/blob/36c1c051b8106371d37f8a747cea0e369a8e5c84/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfiguration.java#L40) I know @leszko did some changes in Spring Boot in the past.

**Comment by jerrinot** [comment]
to reproduce it just create a new spring boot project, add Jet into pom.xml, add this into your Configuration class:
```java
    @Bean
    public JetInstance jet() {
        return Jet.newJetInstance();
    }
```
and add hazelcast.xml to your classpath.

**Comment by nfrankel** [comment]
If you create your own `HazelcastInstance`, this will work. The auto-configuration class is run only if it doesn't find one:

```
@ConditionalOnMissingBean(HazelcastInstance.class)
class HazelcastServerConfiguration
```

(see https://github.com/spring-projects/spring-boot/blob/a66c4d3096877b33119366ea9ea165c892b462c8/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastServerConfiguration.java)

**Comment by jerrinot** [comment]
I do not want `HazelcastInstance` bean at all. I want `JetInstance` bean.

**Comment by jerrinot** [comment]
a workaround it to extract the embedded `HazelcastInstance` from `JetInstance`:
```java
    @Bean
    public HazelcastInstance imdg(JetInstance jet) {
        return jet.getHazelcastInstance();
    }
```

**Comment by nfrankel** [comment]
> The problem is that with Jet you automatically have all IMDG classes available and this confuses the auto-configurator.

Exactly. And Spring Boot doesn't know about Jet...

**Comment by jerrinot** [comment]
this should do the trick:
```
✔ ~/devel/oss/spring-boot [master|✔] 
23:07 $ git diff
diff --git a/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfiguration.java b/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfiguration.java
index 353a39fa9c..43a3e91799 100644
--- a/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfiguration.java
+++ b/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfiguration.java
@@ -20,6 +20,7 @@ import com.hazelcast.core.HazelcastInstance;
 
 import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingClass;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.Import;
@@ -36,6 +37,7 @@ import org.springframework.context.annotation.Import;
  */
 @Configuration(proxyBeanMethods = false)
 @ConditionalOnClass(HazelcastInstance.class)
+@ConditionalOnMissingClass("com.hazelcast.jet.JetInstance")
 @EnableConfigurationProperties(HazelcastProperties.class)
 @Import({ HazelcastClientConfiguration.class, HazelcastServerConfiguration.class })
 public class HazelcastAutoConfiguration {
✔ ~/devel/oss/spring-boot [master|✚ 1] 
```

Now the hardest part: Figure out how to write a test on top of the existing Spring Boot test infrastructure.

**Comment by snicoll** [comment]
> but you probably want to start Jet instance.

There's no way for us to know that. The fact that Jet wraps an `HazelcastInstance` rather than reusing one (if I understood things correctly) is making it very hard for Spring Boot to do the right thing. 

> I do not want HazelcastInstance bean at all. I want JetInstance bean.

What is the problem having a `HazelcastInstance`?

**Comment by nfrankel** [comment]
Welcome @snicoll 🤗

> What is the problem if having a HazelcastInstance bean?

It's not necessary when running Jet as Jet embeds its own, and it will consume additional resources for nothing.

**Comment by snicoll** [comment]
Yo. 

I didn't mean to have two but rather than Jet would use an existing one rather than embedding its own. Given the bootstrap logic of `HazelcastInstance` that feels quite consistent to me to at least offer that option.

**Comment by nfrankel** [comment]
Or... we could give you the context why it is like it is, and see how to work together to offer a Jet Spring Boot starter.

**Comment by snicoll** [comment]
The point I was trying to make is that we have to keep a reasonable amount of complexity on our side. Offering support for Jet in Spring Boot is obviously going to fix this issue but I'd like us to discuss how we can fulfil this use case without having to do so.

If we look at the original description of this issue, there is a problem IMO.

> Spring Boot automatically creates a Hazelcast instance when either Config bean exist or when it finds a configuration file.
>
> This makes Hazelcast Jet hard to use in the Spring Boot container: Spring will try to create a new IMDG instance when it detects hazelcast.xml (or any other config file), but you probably want to start Jet instance.

Creating an `HazelcastInstance` when `hazelcast.xml` is on the classpath looks consistent with [Hazelcast's own behaviour](https://github.com/hazelcast/hazelcast/blob/895874b6c1cf497d5f4dfb5e31aace4bd59a8033/hazelcast/src/main/java/com/hazelcast/core/Hazelcast.java#L110-L124). I understand this it is the wrong decision when one decides to use Jet. I am arguing that this sounds a bit complex and maybe it doesn't have to be the way. 

If creating an `HazelcastInstance` automatically is not the sensible thing to do (anymore), then I am happy to revisit that part of the auto-configuration provided that there is some clarification on your side about that.

**Comment by jerrinot** [comment]
Hello @snicoll, thank you for taking time to think about it and providing your feedback. 

I thought I would share a bit of background: 
There are 2 different projects: Hazelcast IMDG and Hazelcast Jet. From purely a technically perspective Jet is a superset of IMDG, but they solve a different problem. 

The main entry point to IMDG is `Hazelcast.newHazelcastInstance()` - this starts a new IMDG instance.

The main entry point to Jet is `Jet.newJetInstance()` - this starts a new Jet instance.

Given Jet is a superset of IMDG then the Jet JAR also contains all classes from IMDG, including `Hazelcast.newHazelcastInstance()`. However you almost never want to directly start a new IMDG instance when you use Jet, perhaps expect in situations when you are in the process of migrating from our old MapReduce impl to Jet or something like this. 

Here is my idea of the ideal behaviour:
1. Spring Boot creates a new `HazelcastInstance` when IMDG JAR is on a classpath
2. Spring Boot creates a new `JetInstance` when Jet JAR is on a classpath

This creates a consistent and predictable behaviour for Spring Boot users. 

Some time ago @neilstevenson prepared a [change set](https://github.com/spring-projects/spring-boot/pull/13317) introducing the #2, but it was rejected. I can understand why: Jet adoption is lower than IMDG, accepting new code increases your maintenance burden and you have to be very picky when committing to support a new project. So my idea with the [Pull Request I sent today](https://github.com/spring-projects/spring-boot/pull/19483)  was that Spring Boot would stop creating `HazelcastInstance` when it's actually Jet JAR on a classpath and we would include Spring Boot auto-configuration support in our own Jet  - Spring integration module. When the module matures and Jet adoption increase (and it will, mark my words;)) and the Spring Boot project is happy to consider including a first class Jet integration then we will be happy to donate the integration code.

**Comment by neilstevenson** [comment]
I still have the code for that PR on archive. Happy to revive, refresh and resubmit, will solve this issue. Just let me know.

**Comment by gurbuzali** [comment]
It'll be fixed in Spring Boot 2.3.0
See https://github.com/spring-projects/spring-boot/pull/20729
